dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java

```diff
@@ -32,77 +32,96 @@
-	
-	public TokenAwareSelection() {
-		
-		this.tokenMapper = new BinarySearchTokenMapper(new Murmur1HashPartitioner());
-	}
-
-	@Override
-	public void initWithHosts(Map<HostToken, HostConnectionPool<CL>> hPools) {
-		
-		tokenPools.putAll(CollectionUtils.transformMapKeys(hPools, new Transform<HostToken, Long>() {
-
-			@Override
-			public Long get(HostToken x) {
-				return x.getToken();
-			}
-			
-		}));
-
-		this.tokenMapper.initSearchMecahnism(hPools.keySet());
-	}
-
-	@Override
-	public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op) throws NoAvailableHostsException {
-		
-		String key = op.getKey();
-		HostToken hToken = this.getTokenForKey(key);
-		
-		HostConnectionPool<CL> hostPool = null;
-		if (hToken != null) {
-			hostPool = tokenPools.get(hToken.getToken());
-		}
-		
-		if (hostPool == null) {
-			throw new NoAvailableHostsException("Could not find host connection pool for key: " + key + ", hash: " +
-                    tokenMapper.hash(key));
-		}
-		
-		return hostPool;
-	}
-
-	@Override
-	public Map<HostConnectionPool<CL>,BaseOperation<CL,?>> getPoolsForOperationBatch(Collection<BaseOperation<CL, ?>> ops) throws NoAvailableHostsException {
-		throw new RuntimeException("Not Implemented");
-	}
-	
-	@Override
-	public List<HostConnectionPool<CL>> getOrderedHostPools() {
-		return new ArrayList<HostConnectionPool<CL>>(tokenPools.values());
-	}
-	
-	@Override
-	public HostConnectionPool<CL> getPoolForToken(Long token) {
-		return tokenPools.get(token);
-	}
-	
-	public List<HostConnectionPool<CL>> getPoolsForTokens(Long start, Long end) {
-		throw new RuntimeException("Not Implemented");
-	}
+    private final BinarySearchTokenMapper tokenMapper;
+
+    private final ConcurrentHashMap<Long, HostConnectionPool<CL>> tokenPools = new ConcurrentHashMap<Long, HostConnectionPool<CL>>();
+
+    public TokenAwareSelection() {
+
+        this.tokenMapper = new BinarySearchTokenMapper(new Murmur1HashPartitioner());
+    }
+
+    @Override
+    public void initWithHosts(Map<HostToken, HostConnectionPool<CL>> hPools) {
+
+        tokenPools.putAll(CollectionUtils.transformMapKeys(hPools, new Transform<HostToken, Long>() {
+
+            @Override
+            public Long get(HostToken x) {
+                return x.getToken();
+            }
+
+        }));
+
+        this.tokenMapper.initSearchMecahnism(hPools.keySet());
+    }
+
+    /**
+     * If a hashtag has been given then we give priority to it to select to the node.
+     * Otherwise, we use the key.
+     */
+    @Override
+    public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op) throws NoAvailableHostsException {
+
+        String key = op.getKey();
+        String hashtag = op.getHashtag();
+
+        HostToken hToken = null;
+        if (hashtag == null) {
+            hToken = this.getTokenForKey(key);
+        } else {
+            hToken = this.getTokenForKey(hashtag);
```

Comment:
How will this work? From what I undestand hashtag is something like this "{}",
you should find out the content of the hashed key and use it.